Add 'cost' and 'margin' to stake pools

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#1090 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have added cost and margin to 'PoolRegistrationCertificate'  and adjusted Jörmungandr binary decoders to also parse these.
- [x] I have slightly revised the pool database layer to not store (resp. read) only pool owners, but store (resp. read) registration certificates entirely.
- [x] I have adjusted the logic from `listStakePools` to attach `cost` and `margin` to each pool 
- [x] I have extended the API types
- [x] adjusted the swagger file to include these 


# Comments

<!-- Additional comments or screenshots to attach if any -->

```
Cardano.Pool.DB.Sqlite
  Sqlite
    Stake Pool properties
      putPoolRegistration then readPoolRegistration yields expected result
        +++ OK, passed 100 tests.
      rollback of pool Registrations
        +++ OK, passed 100 tests:
        87% rolled back some
        77% owner has many pools
```

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
add cost and margin to 'PoolRegistrationCertificate'

KtorZ committed:
store pool registration certificates on-chain (and not only pool owners)

KtorZ committed:
add cost and margin to stake pools in the API

KtorZ committed:
update swagger schema and JSON api golden to include cost and margin

paweljakubas commented on lib/core/src/Cardano/Pool/DB.hs:
```diff
@@ -73,23 +78,23 @@ data DBLayer m = forall stm. MonadFail stm => DBLayer
         -> stm ()
-        -- ^ Add a mapping between stake pools and owners. If the mapping
-        -- already exists, this will be a no-op.
+        -- ^ Add a mapping between stake pools and their certificate. If the
```
**[suggestion]**

```suggestion
        -- ^ Add a mapping between stake pools and their corresponding certificates. If the
```

paweljakubas commented on lib/core/src/Cardano/Pool/DB/Model.hs:
```diff
@@ -85,6 +92,9 @@ data PoolDatabase = PoolDatabase
     -- ^ Mapping between pool ids and owners
 
+    , metadata :: !(Map PoolId (Percentage, Quantity "lovelace" Word64))
+    -- ^ On-chain metadata associated with pools
```
:+1:

paweljakubas commented on lib/core/src/Cardano/Pool/DB/Model.hs:
```diff
@@ -156,18 +166,35 @@ mReadStakeDistribution epoch db@PoolDatabase{distributions} =
+mPutPoolRegistration registration db@PoolDatabase{owners,metadata} =
     ( Right ()
-    , db { owners = Set.insert entry owners }
+    , db { owners = Set.union entries owners
```
:+1:

paweljakubas commented on lib/core/src/Cardano/Pool/DB/Model.hs:
```diff
@@ -156,18 +166,35 @@ mReadStakeDistribution epoch db@PoolDatabase{distributions} =
+    ( Right $ case Map.lookup poolId metadata of
+        Nothing -> Nothing
+        Just (poolMargin, poolCost) ->
+            let poolOwners = L.sortOn toText
```
**[minor question]**

So we are sorting here on name? If yes, why not based on margin or cost?

KtorZ replied:
This is sorting the _owners_ inside **each** pool, not the pool themselves. This is really just to avoid the database (and the model) returning the list of owner in arbitrary order. This way, we can expect successive calls to return the exact same thing.

paweljakubas commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -164,15 +171,33 @@ newDBLayer logConfig trace fp = do
             repsert
+                (PoolMetadataKey poolId)
+                (PoolMetadata poolId poolMargin_ poolCost_)
+            forM_ poolOwners $ \poolOwner -> repsert
```
**[question]**

`repsert` is used here. so "inserting" many rows at once and (a) if no entry then insert, (b) if entry present update it. Is it ok to have already present `poolId` and new `poolOwner` value? If no then shouldn't be there error? Moreover, don't we afraid of this `repsert` because of performance? So, we are updating many not changed rows every call?

KtorZ replied:
Hmmmm... So, there's nothing _wrong_ with replacing the value. However, thinking again about this well, it would probably be better to simply use `insert_` and make sure that certificates are rolled back correctly like the rest (which isn't the case right now, they are really just treated as "static" metadata, hence the repsert).

paweljakubas commented on lib/core/src/Cardano/Pool/DB/Sqlite/TH.hs:
```diff
@@ -76,4 +76,13 @@ PoolOwner sql=pool_owner
     deriving Show Generic
+
+-- Mapping of metadata to pool
+PoolMetadata sql=pool_metadata
```
**[remark]**

good idea to decouple PoolMetadata from PoolOwner. The former one is expected to have different update rate. And here I am thinking once again about repsert comment from above :thinking:

KtorZ replied:
They are actually updated at the very same time (when a registration certificate is discover on-chain). The reason they are decoupled is because there's a many:many relationship between pools and owners (one pool can have multiple owners, and an owner multiple pools).

paweljakubas commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -351,31 +358,28 @@ prop_putStakePutStake DBLayer {..} epoch a b =
+    -> [PoolRegistrationCertificate]
     -> Property
-prop_stakePoolOwner DBLayer {..} (StakePoolOwnersFixture poolId entries) =
+prop_poolRegistration DBLayer {..} entries =
```
**[remark]**

As we have now  PoolMetadata decoupled from PoolOwner it would be nice to introduce properties like : 
1. registering pools and see if we ONLY have pool metadata relating to this
2. registering pools and then we have pool metadata not belonging to the pool we registered - and what happens?
3. registering and deregistring some pools and what about pool metadata consistency (maybe we cannot deregister pool)

KtorZ replied:
We can't retire pool at the moment indeed, we can only register new pools indeed. The property currently test that for any arbitrary list of registration, if you insert them, you can read them back, therefore making sure that the binding between owners and certificate is correctly respected. It covers your second case (where many metadata does not belong to your pool, but still you don't get extra metadata when fetching) and your first case as well (when fetching, you only get your metadata).

paweljakubas commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -276,7 +279,7 @@ data Fragment
     | StakeDelegation (StakeDelegationType, ChimericAccount, Tx)
     -- ^ A signed transaction with stake pool delegation
-    | PoolRegistration (PoolId, [PoolOwner], Tx)
+    | PoolRegistration (PoolId, [PoolOwner], TaxParameters, Tx)
```
:+1:

paweljakubas reviewed (approved):
I like very much that PoolMetadata is decoupled from PoolOwner. Left some questions/remarks flying mainly around this idea. And also property tests suggestions to make sure those two tables are consistent as I think they should be

KtorZ committed:
properly track registration with rollbacks too

- Use insert_ and insertMany_ instead of repsert
- Added unit test for checking that registration are rolledback correctly
- Rename 'PoolMetadata' table to 'PoolRegistration' for clarity

KtorZ committed:
Add extra checks to STAKE_POOLS_LIST_01 for cost and margin

KtorZ committed:
fix DB model for pools registration

KtorZ committed:
extend pool DB layer and stake pool layer to allow listing known pools

KtorZ committed:
add small property to check that pool ids are listed correctly

KtorZ committed:
do not attempt to re-store known registration from genesis block when already stored

KtorZ committed:
use 'knownStakePools' in the Api server instead of the heavy 'listStakePools'
